### PR TITLE
Optimizations for the WritePdbInfo function

### PR DESCRIPTION
**Pull request by jack-pappas**
I made some basic optimizations to the ``WritePdbInfo`` function in an attempt to speed it up; mostly, I've removed some closure allocations (using ``for`` loops instead of ``.iter`` functions) and changed an unnecessary use of a ref cell to use a normal variable binding instead.

Overall compilation performance seems to be only slightly improved, though I've only tested on my development machine. The benefit may be greater for machines with slower CPU and/or memory (e.g., CI servers running on VMs).

### CODE REVIEW

**Comment by latkin** [comment]
Thanks Jack.  Do you have any reproducible measurements/tests that show the benefit of these changes?

**Comment by dsyme** [comment]
@jack-pappas  - yes, this would need perf figures

@latkin and I may need to write up some notes about what constitutes a minimum-bar for a compiler perf improvement to be accepted - I think minimally

- Reliable, reproducible > 1% perf improvement. See also my other comment about automated CI compiler perf testing - if we could get that in place then we'd see the effect more easily.

- Very, very clear there is absolutely no chance of a regression in logic

Without the demonstrated perf improvement a checkin couldn't be accepted, hence the need for CI to make it cheap to reproduce these.

BTW I notice some of the code changes are just cleanup - you could resubmit those as a separate checkin labelled "cleanup" - it is very, very helpful to separate code cleanup from perf improvements.  Again @latkin and I may need to write notes on what constitutes an acceptable "cleanup" checkin and at what stage.

File: src/absil/ilwrite.fs [hunk 1]
```diff
@@ -223,17 +224,19 @@ type PdbData =
...
+        with _ ->
+            error(Error(FSComp.SR.ilwriteErrorCreatingPdb(fpdb), rangeCmdArgs))
 
     match info.EntryPoint with 
     | None -> () 
-    | Some x -> pdbSetUserEntryPoint !pdbw x 
+    | Some x -> pdbSetUserEntryPoint pdbw x
 
-    let docs = info.Documents |> Array.map (fun doc -> pdbDefineDocument !pdbw doc.File)
+    //assert(info.Documents |> Array.forall (fun doc -> System.IO.File.Exists doc.File))
```
Comment (enricosada) [inline src/absil/ilwrite.fs]: commented code

File: src/absil/ilwrite.fs [hunk 2]
```diff
@@ -254,83 +257,102 @@ let WritePdbInfo fixupOverlappingSequencePoints showTimes f fpdb info =
...
+                KeyValuePair (n, sp)
+        allSps |> Array.sortInPlaceBy (fun kvp -> kvp.Key)
+
+    let mutable spOffset = 0
+    let infoMethods = info.Methods
+    for i = 0 to infoMethods.Length - 1 do
+          let minfo = infoMethods.[i]
+
+          let sps = System.ArraySegment (allSps, spOffset, spCounts.[i])
+          //let sps = Array.sub allSps spOffset spCounts.[i]
```
Comment (enricosada) [inline src/absil/ilwrite.fs]: commented code

File: src/absil/ilwrite.fs [hunk 3]
```diff
@@ -3439,6 +3461,8 @@ let writeILMetadataAndCode (generatePdb,desiredMetadataVersion,ilg,emitTailcalls
     let strings,userStrings,blobs,guids,tables,entryPointToken,code,requiredStringFixups,data,resources,pdbData,mappings = 
       generateIL requiredDataFixups (desiredMetadataVersion,generatePdb,ilg,emitTailcalls,showTimes) modul noDebugData cilStartAddress
 
+    //assert(pdbData.Documents |> Array.forall (fun doc -> System.IO.File.Exists doc.File))
```
Comment (enricosada) [inline src/absil/ilwrite.fs]: commented code

File: src/absil/ilwrite.fs [hunk 4]
```diff
@@ -3996,6 +4020,8 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
           let entryPointToken,code,codePadding,metadata,data,resources,requiredDataFixups,pdbData,mappings = 
             writeILMetadataAndCode ((pdbfile <> None), desiredMetadataVersion, ilg,emitTailcalls,showTimes) modul noDebugData next
 
+          //assert(pdbData.Documents |> Array.forall (fun doc -> System.IO.File.Exists doc.File))
```
Comment (enricosada) [inline src/absil/ilwrite.fs]: commented code

**Comment by latkin** [comment]
Thank you for the submission. Per the [contribution guidelines](https://github.com/Microsoft/visualfsharp/blob/fsharp4/CONTRIBUTING.md#mimimum-bar-for-performance-improvement-pull-requests), perf optimizations should include tests and quantitative analysis showing reliable gains from the change, which others can reproduce.  Please re-open when tests and analysis are in place.

**Pull request closed without merging**